### ISSUE TITLE
Run ASRT on smartphones.

### ISSUE THREAD

**Issue by Evanston0624** [nl8590687/ASRT_SpeechRecognition#328]
我希望在android & IOS上運行ASRT，我使用python進行後續的測試
首先我將音檔讀入是使用:  
```
import librosa
wav_signal, sample_rate = librosa.load(audio_path, sr=None)
```
接下來我讀取ASRT的模型參數(這個模型除了原始數據外，還加入了CV的TW數據。    
```
from model_zoo.speech_model.keras_backend import SpeechModel251BN
def load_tf_model(model_path):
	AUDIO_LENGTH = 1600
	AUDIO_FEATURE_LENGTH = 200
	CHANNELS = 1
	# 原始拼音=1427、cv-TW=3、空白=1
	OUTPUT_SIZE = 1431
	sm251bn = SpeechModel251BN(
		input_shape=(AUDIO_LENGTH, AUDIO_FEATURE_LENGTH, CHANNELS),
		output_size=OUTPUT_SIZE
    )
	sm251bn.load_weights('./save_models/SpeechModel251bn/SpeechModel251bn_epoch40.model.h5')
	trained_model, base_model = sm251bn.get_model()
	return trained_model, base_model
```

我透過上述的代碼取得欲訓練的模型，trained_model是包含CTC loss的，因此我使用base_model 進行轉換。我共使用了ONNX與TF_lite進行測試:
TF_lite:  
```
def convert_tf_lite(tf_model, save_path):
	# 轉換为 TensorFlow Lite 模型
	converter = tf.lite.TFLiteConverter.from_keras_model(tf_model)
	tflite_model = converter.convert()
	# 保存 TensorFlow Lite 模型
	with open(save_path, 'wb') as f:
		f.write(tflite_model)
	return os.path.isfile(save_path)
```

ONNX:  
```
def convert_tf_onnx(tf_model, save_path, opset):
	import tensorflow as tf
	import tf2onnx
	# 轉換為 ONNX 格式
	onnx_model, _ = tf2onnx.convert.from_keras(tf_model, opset=opset)

	# 保存 ONNX 模型
	with open(save_path, 'wb') as f:
		f.write(onnx_model.SerializeToString())
	return os.path.isfile(save_path)
```
接下來我使用了修改過的Spectrogram進行特徵提取
```
from speech_features import Spectrogram
data_pre = Spectrogram()
audio_features = data_pre.onnx_run(wavsignal=wav_signal, fs=sample_rate)
audio_features = adaptive_padding(input_data=audio_features, target_length=1600)
```
我在原始的Spectrogram類參考run創建了onnx_run，實際上只是匹配輸入參數的維度等資訊。
```
def onnx_run(self, wavsignal, fs=16000):
	if fs != 16000:
		raise ValueError(
			f"[Error] ASRT currently only supports wav audio files with a sampling rate of 16000 Hz, but this "
			f"audio is {fs} Hz.")

	# wav波形 加时间窗以及时移10ms
	time_window = 25  # 单位ms
	window_length = int(fs / 1000 * time_window)  # 计算窗长度的公式，目前全部为400固定值

	wav_arr = np.array(wavsignal)

	range0_end = int(len(wavsignal) / fs * 1000 - time_window) // 10 + 1  # 计算循环终止的位置，也就是最终生成的窗数
	data_input = np.zeros((range0_end, window_length // 2), dtype=np.float64)  # 用于存放最终的频率特征数据
	data_line = np.zeros((1, window_length), dtype=np.float64)

	for i in range(0, range0_end):
		p_start = i * 160
		p_end = p_start + 400

		data_line = wav_arr[p_start:p_end]
		data_line = data_line * self.w  # 加窗
		data_line = np.abs(fft(data_line))

		data_input[i] = data_line[0: window_length // 2]  # 设置为400除以2的值（即200）是取一半数据，因为是对称的

	data_input = np.log(data_input + 1)
	return data_input
```
接下來透過adaptive_padding將輸入的特徵轉換成跟原始輸入相同的尺寸
```
def adaptive_padding(input_data, target_length=1600):
	input_data = input_data.astype(np.float32)

	input_data = np.expand_dims(input_data, axis=0)  # 添加批量维度
	input_data = np.expand_dims(input_data, axis=-1)  # 添加通道维度
	# 计算需要填充的长度
	current_length = input_data.shape[1]
	padding_length = max(0, target_length - current_length)

	# 计算填充宽度
	left_padding = padding_length // 2
	right_padding = padding_length - left_padding
	pad_width = [(0, 0), (left_padding, right_padding), (0, 0), (0, 0)]

	# 进行填充
	padded_data = np.pad(input_data, pad_width, mode='constant').astype(np.float32)

 	return padded_data
```
經由上述轉換後的模型輸出結果，都是空白塊分數最高，後續調用tf.nn.ctc_beam_search_decoder與K.ctc_decode就沒有意義了。

想請問是否有相關的研究或實踐方法可以推薦?又或者我需要提供更多的測試或特定文件?

感謝

**Comment by Evanston0624** [comment]
補充:
相同的模型使用predict_speech_file.py是可以正確預測輸出的
```
import os

from speech_model import ModelSpeech
from model_zoo.speech_model.keras_backend import SpeechModel251BN
from speech_features import Spectrogram
from language_model3 import ModelLanguage

os.environ["CUDA_VISIBLE_DEVICES"] = ""

AUDIO_LENGTH = 1600
AUDIO_FEATURE_LENGTH = 200
CHANNELS = 1
# 默认输出的拼音的表示大小是1428，即1427个拼音+1个空白块
OUTPUT_SIZE = 1431
sm251bn = SpeechModel251BN(
    input_shape=(AUDIO_LENGTH, AUDIO_FEATURE_LENGTH, CHANNELS),
    output_size=OUTPUT_SIZE
)
feat = Spectrogram()
ms = ModelSpeech(sm251bn, feat, max_label_length=64)
now_path = os.path.abspath(os.getcwd())

ms.load_model(now_path+'/save_models/SpeechModel251bn_cv/' + 'SpeechModel251bn_epoch40.model.base.h5')

res = ms.recognize_speech_from_file('test1.wav')
print('*[提示] 声学模型语音识别结果：\n', res)
```

**Comment by nl8590687** [comment]
不建议直接在手机端运行，否则计算性能和依赖环境的安装配置都较为复杂，最佳方案是模型部署于服务器，手机通过API接口调用。具体讲解可以看AI柠檬博客相关文章。

**Comment by Evanston0624** [comment]
您好，我們的服務器在多用戶調用時的響應速度與不如預期，後續我有自己編寫一套透過socket的TCP+UDP實踐註冊與傳遞語音包的多進程程序，但在多用戶組的情況下響應一樣不快。(上述問題可能是存在我們的硬體配置)

因此我想透過ONNX與TF-Lite來實現移動設備推理，我剛剛實際測試已經可以生成結果了，稍晚會把代碼發上來(python的測試代碼)。後續應該會使用java開發app程序，那這部分的工作應該如下:

- [ ] 讀取音檔
- [ ] 轉頻譜
- [ ] onnx模型推理
- [ ] ctc推理

我認為如果可以確定python的數據格式與java上的差異，應該可以正確運行

**Comment by nl8590687** [comment]
单进程因为只有一个计算图资源，多用户并发调用响应速度慢是很正常的，你需要做的是多实例集群部署，负载均衡，而不是单纯的改通信协议。AI模型部署本身就是很耗费资源的。

**Comment by Evanston0624** [comment]
main Activity 代碼:
```
import android.content.Context;
import android.content.res.AssetManager;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import java.io.IOException;
import java.io.InputStream;


public class MainActivity extends AppCompatActivity {
    private static final String TAG = "Main";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // 檢查 test.wav 檔案是否存在
        boolean fileExists = isFileExists(this, R.raw.test);

        if (fileExists) {
            // 如果檔案存在，執行您的相應邏輯
            Toast.makeText(this, "test.wav 檔案存在", Toast.LENGTH_SHORT).show();

            // 讀取並處理音訊檔案
            double[] audioData = DataLoader.readAsInt(this, R.raw.test);

            if (audioData != null && audioData.length > 0) {
                Toast.makeText(this, "audioData 長度：" + audioData.length, Toast.LENGTH_SHORT).show();

                // 處理音訊數據
                Spectrogram spectrogram = new Spectrogram(16000, 25, 10);
                float[][] result = spectrogram.calculateSpectrogram(audioData);

                // padding
                float[][][][] mdl_input = DataLoader.prepareInput(result);

                // check model
                // 创建 AssetManager 实例
                AssetManager assetManager = getAssets();
                // 模型文件路径
                String modelPath = "model/model.tflite";

                // 检查模型文件是否存在
                try {
                    // 打开模型文件的输入流
                    InputStream inputStream = assetManager.open(modelPath);
                    Log.d(TAG, "模型文件存在");
                    // 关闭输入流
                    inputStream.close();
                } catch (IOException e) {
                    Log.d(TAG, "模型文件不存在");
                    e.printStackTrace();
                }

                TfliteInfer tfliteInfer = new TfliteInfer();
                tfliteInfer.loadModel(assetManager, modelPath);
                float [][] mdl_output = tfliteInfer.runInference(mdl_input);

                String str2 = " ka1 qi3 tog1 zi1 yan3 mian4";

                int distance = LevenshteinDistance.computeLevenshteinDistance(ctc_output, str2);
                double similarity = LevenshteinDistance.similarity(ctc_output, str2);
            } else {
                Toast.makeText(this, "audioData 回傳錯誤" , Toast.LENGTH_SHORT).show();
                // 處理讀取失敗的情況
            }
        } else {
            // 如果檔案不存在，執行相應的錯誤處理邏輯
            Toast.makeText(this, "test.wav 檔案不存在", Toast.LENGTH_SHORT).show();
        }
    }
    // 檢查檔案是否存在的方法
    public boolean isFileExists(Context context, int resourceId) {
        try {
            context.getResources().openRawResourceFd(resourceId).close();
            return true;
        } catch (Exception e) {
            return false;
        }
    }
}
```

**Comment by Evanston0624** [comment]
DataLoader，用於載入音訊數據，將音訊轉為float[][][][]，後續在輸入tf_lite時會轉為byte_buffer。
為了確保輸出結果跟python 上運行ASRT一樣，對於此部分的輸出結果有檢查過:shape、max、min，都跟python上運行相同。
此步驟可以優化成:
將PCM-16bit轉為float(4byte)。

```
import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.media.MediaExtractor;
import android.media.MediaFormat;
import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

public class DataLoader {
    private static final String TAG = "DataLoader";

    public static double[] readAsInt(Context context, int resourceId) {
        MediaExtractor extractor = new MediaExtractor();
        try {
            AssetFileDescriptor assetFileDescriptor = context.getResources().openRawResourceFd(resourceId);
            extractor.setDataSource(assetFileDescriptor.getFileDescriptor(), assetFileDescriptor.getStartOffset(), assetFileDescriptor.getLength());

            assetFileDescriptor.close();
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }

        float numSamples = getNumSamples(extractor);
        ByteBuffer audioData = readData(extractor, numSamples);
        Log.d(TAG, "音訊數據的剩餘樣本數量：" + audioData.remaining()/2);

        return convertToIntegerArray(audioData);
    }

    private static int getNumSamples(MediaExtractor extractor) {
        MediaFormat format = extractor.getTrackFormat(0);
        int sampleRate = format.getInteger(MediaFormat.KEY_SAMPLE_RATE);
        int durationUs = (int) format.getLong(MediaFormat.KEY_DURATION);
        Log.d(TAG, "Sample rate: " + sampleRate);
        Log.d(TAG, "Duration (us): " + durationUs);
        durationUs = durationUs/1000;
        sampleRate = sampleRate/1000;
        return durationUs*sampleRate;
    }

    private static ByteBuffer readData(MediaExtractor extractor, float numSamples) {
        ByteBuffer buffer = ByteBuffer.allocate((int) numSamples * 2); // 16-bit PCM, 2 bytes per sample

        extractor.selectTrack(0);

        while (true) {
            int bytesRead = extractor.readSampleData(buffer, buffer.position());
            if (bytesRead < 0) {
                break;
            }
            buffer.position(buffer.position() + bytesRead);
            extractor.advance();
        }
        // Reset buffer position to the beginning
        buffer.rewind();
        return buffer;
    }

    private static double[] convertToIntegerArray(ByteBuffer byteBuffer) {
        byteBuffer.rewind();
        byteBuffer.order(ByteOrder.LITTLE_ENDIAN); // assuming little-endian data
        double[] intData = new double[byteBuffer.remaining() / 2];
        Log.d(TAG, "intData-length：" + intData.length);
        for (int i = 0; i < intData.length; i++) {
            double value = (double)byteBuffer.getShort();
            intData[i] = value; // 不進行無符號轉換
        }
        Log.d(TAG, "intData-length：" + intData.length);
        return intData;
    }
    public static float[][][][] prepareInput(float[][] inputData) {
        int batchSize = 1;
        int height = 1600;
        int width = 200;
        int depth = 1;

        float[][][][] inputArray = new float[batchSize][height][width][depth];
        int padding_length = height - inputData.length;
        int left_padding = padding_length / 2;

        int count = 0;
        for (int i = 0; i < height; i++) {
            if (i >= left_padding && i < left_padding+inputData.length) {
                float[] row = inputData[count];
                for (int j = 0; j < width; j++) {
                    inputArray[0][i][j][0] = row[j];
                }
                count += 1;
            }else {
                for (int j = 0; j < width; j++) {
                    inputArray[0][i][j][0] = 0.0f;
                }
            }
        }
        return inputArray;
    }
    public static String[] processFile(Context context, String fileName) {
        List<String> firstElements = new ArrayList<>();

        try (InputStream is = context.getAssets().open(fileName);
             BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {

            String line;
            while ((line = reader.readLine()) != null) {
                String[] elements = line.split("\t");
                if (elements.length > 0) {
                    firstElements.add(elements[0]);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        firstElements.add(" ");
        return firstElements.toArray(new String[0]);
    }
}
```

**Comment by Evanston0624** [comment]
感謝您對於製作ASR服務的意見與提點，我已經實踐了在Android上使用ASRT訓練的模型(轉為TF-lite)進行推理，並實踐了:

1. 數據前處理(載入音訊的格式>頻譜>padding)
2. 載入模型
3. 調用模型取得輸出
4. 編寫了一個ctc_decode(此部分可能跟原始調用Keras的ctc_decoder不同，但我在小樣本下測試結果沒問題)

尚未實踐
- [ ] 將phoneme轉為詞彙
- [ ] 運行效能測試

將ASRT訓練的模型轉為TF-lite的代碼後續會補上

https://github.com/Evanston0624/ASRT_model_Android/tree/main
README 稍晚會創建

**Comment by Evanston0624** [comment]
更新了ASRT模型轉換為onnx、tf-lite與core-ml的代碼:
[子專案](https://github.com/Evanston0624/ASRT_model_Android/tree/main/python_qtz)

這個子專案可以使用python進行onnx、tf-lite的轉換與推理，後續會完善量化評估等資訊。
